numpy/core/src/multiarray/buffer.c

```diff
@@ -248,14 +260,17 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
             /* Insert padding manually */
             if (*offset > new_offset) {
                 PyErr_SetString(PyExc_RuntimeError,
-                                "This should never happen: Invalid offset in "
-                                "buffer format string generation. Please "
-                                "report a bug to the Numpy developers.");
+                                "The buffer interface does not support "
+                                "overlapping fields");
```

Comment:
Just a thought (weeks later),

A more general way of describing a structured type is as a list of `(name, type, offset)` tuples. This is how, for example, MPI serializes structs (see [here](http://beige.ucs.indiana.edu/I590/node100.html)). It is also how numpy already stores structured types internally.

Since a purpose of PEP3118 is serialization, maybe it would have made sense to specify structured types this way. Eg we could have had something like `'T24{i,0,f0:f,8,f1}` to represent `dtype({'formats': ['i', 'f'], 'offsets': [0, 8], 'names': ['f0', 'f1'], 'itemsize': 24})`.

Since numpy is the main user of the PEP3118 interface, and because the PEP seems to be only partly implemented in cpython currently, there is probably still opportunity for us to work out the details or add to the specification, with some effort.